File: practices/webflux/src/main/java/com/webflux/controller/UserController.java
package com.webflux.controller;

import com.webflux.controller.dto.ProfileImageResponse;
import com.webflux.controller.dto.UserResponse;
import com.webflux.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Mono;

/**
 * packageName    : com.webflux.controller
 * fileName       : UserController
 * author         : okdori
 * date           : 11/27/23
 * description    :
 */

@RequiredArgsConstructor
@RequestMapping("/api/users")
@RestController
public class UserController {
    private final UserService userService;

    @GetMapping("/{userId}")
    public Mono<UserResponse> getUserById(
            @PathVariable String userId
    ) {
        return ReactiveSecurityContextHolder
                .getContext()
                .flatMap(context -> {
                    String name = context.getAuthentication().getName();

                    if (!name.equals(userId)) {
                        return Mono.error(new ResponseStatusException(HttpStatus.UNAUTHORIZED));
                    }

                    return userService.findById(userId)
                            .map(user -> {
                                return new UserResponse(
                                        user.getId(),
                                        user.getName(),
                                        user.getAge(),
                                        user.getFollowCount(),
                                        user.getProfileImage().map(image ->
                                            new ProfileImageResponse(
                                                    image.getId(),
                                                    image.getName(),
                                                    image.getUrl()
                                            ))
                                );
                            }).switchIfEmpty(
                                    Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND))
                            );
                });
    }
}

File: practices/coroutine-r2dbc/src/main/kotlin/com/coroutiner2dbc/controller/UserController.kt
package com.coroutiner2dbc.controller

import com.coroutiner2dbc.common.Image
import com.coroutiner2dbc.common.User
import com.coroutiner2dbc.controller.dto.ProfileImageResponse
import com.coroutiner2dbc.controller.dto.SignupUserRequest
import com.coroutiner2dbc.controller.dto.UserResponse
import com.coroutiner2dbc.service.UserCoroutineService
import kotlinx.coroutines.reactor.awaitSingle
import org.springframework.http.HttpStatus
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException

/**
 * packageName    : com.coroutiner2dbc.controller
 * fileName       : UserController
 * author         : okdori
 * date           : 3/8/24
 * description    :
 */

@RequestMapping("/api/users")
@RestController
class UserController(
    private val userService: UserCoroutineService,
) {
    @GetMapping("/{userId}")
    suspend fun getUserById(
        @PathVariable userId: String
    ): UserResponse {
        val context = ReactiveSecurityContextHolder
            .getContext()
            .awaitSingle()

        val name = context.authentication.name
        if (name != userId) {
            throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
        }

        return userService.findById(userId)
            ?.let(this::map)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/signup")
    suspend fun signupUser(
        @RequestBody request: SignupUserRequest
    ): UserResponse {
        return userService.createUser(
            request.name, request.age,
            request.password, request.profileImageId
        ).let(this::map)
    }

    private fun map(user: User): UserResponse {
        return UserResponse(
            user.id,
            user.name,
            user.age,
            user.followCount,
            user.profileImage.map { image: Image ->
                ProfileImageResponse(
                    image.id,
                    image.name,
                    image.url
                )
            }
        )
    }
}
